src/daft-physical-plan/src/ops/cross_join.rs

```diff
@@ -0,0 +1,88 @@
+use std::sync::Arc;
+
+use common_display::tree::TreeDisplay;
+use daft_core::join::JoinSide;
+use daft_logical_plan::{
+    partitioning::{
+        ClusteringSpecRef, HashClusteringConfig, RangeClusteringConfig, UnknownClusteringConfig,
+    },
+    ClusteringSpec,
+};
+use serde::{Deserialize, Serialize};
+
+use crate::PhysicalPlanRef;
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct CrossJoin {
+    pub left: PhysicalPlanRef,
+    pub right: PhysicalPlanRef,
+
+    /// the side that is used for the outer loop is relevant for maintaining the clustering spec
+    pub outer_loop_side: JoinSide,
+    pub clustering_spec: ClusteringSpecRef,
+}
+
+impl CrossJoin {
+    pub(crate) fn new(left: PhysicalPlanRef, right: PhysicalPlanRef) -> Self {
+        let left_spec = left.clustering_spec();
+        let right_spec = right.clustering_spec();
+
+        let num_partitions = left_spec.num_partitions() * right_spec.num_partitions();
+
+        fn try_clustering_spec_from(
+            spec: &ClusteringSpec,
+            num_partitions: usize,
+        ) -> Option<ClusteringSpec> {
+            if spec.num_partitions() == 1 {
+                match spec {
+                    ClusteringSpec::Hash(HashClusteringConfig { by, .. }) => Some(
+                        ClusteringSpec::Hash(HashClusteringConfig::new(num_partitions, by.clone())),
+                    ),
+                    ClusteringSpec::Range(RangeClusteringConfig { by, descending, .. }) => {
+                        Some(ClusteringSpec::Range(RangeClusteringConfig::new(
+                            num_partitions,
+                            by.clone(),
+                            descending.clone(),
+                        )))
+                    }
+                    _ => None,
+                }
+            } else {
+                None
+            }
+        }
+
+        let (outer_loop_side, clustering_spec) =
+            if let Some(spec) = try_clustering_spec_from(&left_spec, num_partitions) {
+                (JoinSide::Left, spec)
+            } else if let Some(spec) = try_clustering_spec_from(&right_spec, num_partitions) {
```

Comment:
A couple questions here:
1. What if both sides have a clustering spec with `num_partitions = 1`? Currently it looks like we take the left and ignore the right.
2. Is the output clustering spec maintained if the num_partitions increases? E.g. if left has hash clustering and num_partitions ==1, then output has hash clustering and num_partitions == left_partitions * right_partitions, which could be > 1. Now each output partition essentially contains all of left (because it is the outer side), which means there are duplicate values of left across partitions, would it still be considered as clustered by hash?

Reply:
Actually my implementation is wrong. What I would like to do is the following: one of the sides can propagate its hash or range clustering spec only if the _other side_ has one partition. In the case of both sides having one partition, it doesn't particularly matter what partitioning it uses, we just default to the left.